crates/sui-indexer-restorer/src/snapshot.rs

```diff
@@ -0,0 +1,214 @@
+use std::sync::Arc;
+
+use anyhow::Error;
+use diesel_async::RunQueryDsl;
+use futures::future::{AbortHandle, AbortRegistration, Abortable};
+use indicatif::{MultiProgress, ProgressBar, ProgressStyle};
+use object_store::path::Path;
+use sui_indexer_alt::models::objects::{create_stored_obj_info, StoredObjInfo};
+use tokio::sync::{Mutex, Semaphore};
+use tokio::task;
+use tracing::info;
+use url::Url;
+
+use sui_config::object_storage_config::{ObjectStoreConfig, ObjectStoreType};
+use sui_core::authority::authority_store_tables::LiveObject;
+use sui_field_count::FieldCount;
+use sui_indexer_alt::db::{Db, DbArgs};
+use sui_indexer_alt::schema::obj_info;
+use sui_snapshot::{
+    reader::{download_bytes, LiveObjectIter, StateSnapshotReaderV1},
+    FileMetadata,
+};
+use sui_storage::object_store::ObjectStoreGetExt;
+
+use crate::Args;
+
+pub type DigestByBucketAndPartition = BTreeMap<u32, BTreeMap<u32, [u8; 32]>>;
+
+pub struct SnapshotRestorer {
+    pub restore_args: Args,
+    pub next_checkpoint_after_epoch: u64,
+    pub snapshot_reader: StateSnapshotReaderV1,
+    pub db: Db,
+}
+
+impl SnapshotRestorer {
+    pub async fn new(args: &Args, next_checkpoint_after_epoch: u64) -> Result<Self, Error> {
+        let remote_store_config = ObjectStoreConfig {
+            object_store: Some(ObjectStoreType::S3),
+            aws_endpoint: Some(args.endpoint.clone()),
+            aws_virtual_hosted_style_request: true,
+            object_store_connection_limit: args.concurrency,
+            no_sign_request: true,
+            ..Default::default()
+        };
+
+        let local_path = PathBuf::from(&args.snapshot_local_dir);
+        let snapshot_dir = local_path.join("snapshot");
+        let local_store_config = ObjectStoreConfig {
+            object_store: Some(ObjectStoreType::File),
+            directory: Some(snapshot_dir.clone().to_path_buf()),
+            ..Default::default()
+        };
+
+        let m = MultiProgress::new();
+        let snapshot_reader = StateSnapshotReaderV1::new(
+            args.start_epoch,
+            &remote_store_config,
+            &local_store_config,
+            usize::MAX,
+            NonZeroUsize::new(args.concurrency).unwrap(),
+            m,
+            true, // skip_reset_local_store
+        )
+        .await?;
+
+        let database_url = Url::parse(&args.database_url)?;
+        let db_args = DbArgs::with_url(database_url);
+        let db = Db::new(db_args).await?;
+
+        Ok(Self {
+            restore_args: args.clone(),
+            snapshot_reader,
+            db,
+            next_checkpoint_after_epoch,
+        })
+    }
+
+    pub async fn restore(&mut self) -> Result<(), Error> {
+        info!(
+            "Starting snapshot restore from epoch {}",
+            self.restore_args.start_epoch
+        );
+        let (sha3_digests, num_part_files) = self.snapshot_reader.compute_checksum().await?;
+        let (_abort_handle, abort_registration) = AbortHandle::new_pair();
```

Comment:
It seems like you aren't using the `AbortHandle`, right? I would expect some logic somewhere that listens for an interrupt and triggers the abort handle in that case, but it seems like that side of the pair is unused.

Also, is it safe to abort restoration at any point? It may be better to use a `CancellationToken` and check it regularly while restoring to ensure we stop work at a safe point, and don't corrupt anything. The indexer framework includes a `graceful_shutdown` function which works with cancellation tokens and listens for `Ctrl + C`.

Reply:
> is it safe to abort restoration at any point

yeah if the restore is interrupted, it's not safe to resume and continue so aborting gracefully is not very useful, removing the aborting logic.